src/main/java/minijava/lexer/BasicLexerInput.java

```diff
@@ -0,0 +1,60 @@
+package minijava.lexer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import minijava.MJError;
+
+/** A basic lexer input stream without advanced buffering. */
```

Comment:
Why exactly don't you use buffering? This kills performance. At least wrap the InputStream somewhere in a BufferedInputStream.